### PR TITLE
Adds unitary testing for routed circuits

### PR DESCRIPTION
**Pull request by ammareltigani**
Adds utilitiy functions that tests whether two circuits define the same unitary up to permutation of qubits.

### CODE REVIEW

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
As discussed offline, let's change the API of the testing method (and router implementation) to take an `actual` and `expected` circuit with the same set of qubits, along with a `qubit_map` that maps the initial qubits of `actual` to it's final qubit mapping. As a fallback, if the `qubit_map` is not specified, the method can potentially try to construct a qubit map by looking at the swaps operations in the `actual` circuit.

**Comment by ammareltigani** [comment]
@tanujkhattar I've made the changes we discussed without the fallback strategy as a v1. I can follow up with that with another PR once we have a fully functioning basic routing implementation pushed to Cirq.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/testing/circuit_compare.py [hunk 1]
```diff
@@ -220,6 +220,35 @@ def _first_differing_moment_index(
     return None  # coverage: ignore
 
 
+def assert_circuits_have_same_unitary_given_final_permutation(
+    actual: circuits.Circuit, expected: circuits.Circuit, qubit_map: Dict[ops.Qid, ops.Qid]
```
Comment (tanujkhattar) [inline cirq-core/cirq/testing/circuit_compare.py:224]:
```suggestion
    actual: cirq.AbstractCircuit, expected: cirq.AbstractCircuit, qubit_map: Dict[ops.Qid, ops.Qid]
```

File: cirq-core/cirq/testing/circuit_compare.py [hunk 2]
```diff
@@ -220,6 +220,35 @@ def _first_differing_moment_index(
...
+    if any(not set(qs).issubset(actual.all_qubits()) for qs in zip(*tuple(qubit_map.items()))):
+        raise ValueError(
+            f"The dictionary 'qubit_map' must be a mapping of the qubits in the circuit 'actual'"
+            f" to themselves."
+        )
+
+    inverse_qubit_perm = {v: k for k, v in qubit_map.items()}
+    sorted_qubits, initial_qubits = zip(*sorted(inverse_qubit_perm.items(), key=lambda x: x[0]))
+    inverse_permutation = [sorted_qubits.index(q) for q in initial_qubits]
+    actual.append(ops.QubitPermutationGate(list(inverse_permutation)).on(*sorted_qubits))
```
Comment (tanujkhattar) [inline cirq-core/cirq/testing/circuit_compare.py:245]: We cannot modify the input circuit. mypy should also start complaining once you change the circuit type to `AbstractCircuit`

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

**Review by tanujkhattar — APPROVED** [review]

**Pull request merged**